docs: how to manage remote systems (#547)
canonical/pebble: docs/how-to/manage-remote-systems.md

# How to use Pebble to manage remote systems

> Managing a cluster of servers remotely is no mean task. It is not only time-consuming since servers need regular software updates, configuration changes, file transfers, and command executions, but it also presents security challenges because remote access requires opening ports, which increases the potential attack surface.

Reviewer: I think you mean "no mean feat" :slightly_smiling_face:

I suggest breaking the long sentence - something like this:
```suggestion
Managing a cluster of servers remotely is no mean feat. It is time-consuming because servers need regular software updates, configuration changes, file transfers, and command executions. It also presents security challenges because remote access requires opening ports, which increases the potential attack surface.
```

> While it's possible to manually execute commands on remote servers and transfer files from/to remote servers, it is not the most efficient (or secure) approach. Several other methods exist for remote command execution and system administration. Secure Shell (SSH) is one of them, and there are also dedicated configuration management tools like Ansible. However, to use these tools, an extra port for SSH needs to be opened, which not only increases the operational overhead of managing public/private keys but also the potential attack surface. Remember March 29, 2024 which seemed to be another normal Friday? A developer shocked the world by revealing an XZ Utils (data-compression utilities) backdoor which could potentially enable unauthorized access via SSH and remote code execution (read the full story [here](https://www.ssh.com/blog/a-recap-of-the-openssh-and-xz-liblzma-incident)).
> 
> ---
> 
> ## Pebble in remote systems
> 
> Pebble provides various commands and APIs to run commands and manage files and directories in remote systems. With these capabilities, if we run Pebble in a remote system like in a separate container, we can use Pebble APIs to coordinate with the remote system without opening a new port.

Reviewer: If I'm being ultra-picky, I would leave out the whole paragraph starting "While it's possible to manually execute commands on remote servers", as I don't think it's essential to the purpose of the doc. However, I do appreciate that it's interesting background context.

How about combining all of this and compressing it slightly? Like this:

```suggestion
## Why use Pebble?

Pebble provides commands and a REST API to run commands and manage files and directories in remote systems. If we run Pebble in a remote system, for example in a separate container, we can use Pebble to coordinate with the remote system without opening a new port.

While it's possible to manually run commands on remote servers and transfer files from/to remote servers, this isn't the most efficient or secure approach. Even a configuration management tool such as Ansible requires an extra port to be opened for Secure Shell (SSH), which not only increases the operational overhead of managing public/private keys but also the potential attack surface.

Remember March 29, 2024 which seemed to be another normal Friday? A developer shocked the world by revealing an XZ Utils (data-compression utilities) backdoor which could potentially enable unauthorized access via SSH and remote code execution (read the full story [here](https://www.ssh.com/blog/a-recap-of-the-openssh-and-xz-liblzma-incident)).
```

For the part about the XZ Utils incident, is there an alternative write-up that we could link to? Or perhaps an article that talks more generally (and neutrally) about why opening ports is an increased attack surface? The link to ssh.com looks like a marketing piece for that company.

Author: For the XZ utils incident, how about using a Wikipedia link? https://en.wikipedia.org/wiki/XZ_Utils_backdoor.

---

## Run commands in remote systems

One common task in system administration is updating and installing packages on remote servers. With Pebble, we can use the `pebble exec` command to achieve this.

> For example, if Pebble is running with a user with root privilege, we can use this command to update and install packages in remote systems:

Reviewer: Small suggestion for clarity:
```suggestion
For example, if Pebble is running as user with root privileges, we can use this command to update and install packages in remote systems:
```

```{terminal}
:input: pebble exec apt update
Hit:1 http://ports.ubuntu.com/ubuntu-ports noble InRelease
Get:2 http://ports.ubuntu.com/ubuntu-ports noble-updates InRelease [126 kB]
Get:3 http://ports.ubuntu.com/ubuntu-ports noble-backports InRelease [126 kB]
...
Fetched 3121 kB in 38s (81.5 kB/s)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
41 packages can be upgraded. Run 'apt list --upgradable' to see them.
```

To install a package, run:

```bash
pebble exec apt install cowsay
```

To confirm the package is successfully installed in the remote system, run:

```{terminal}
:input: pebble exec cowsay moo
 _____
< moo >
 -----
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||
```

For more information on the Pebble `exec` command, see {ref}`reference_pebble_exec_command`.

---

## Manage files in remote systems

> Another common task in remote systems is configuration management, like updating configuration files, pushing files to the servers, pulling files generated by running services, and so on.

Reviewer: Minor suggestion for readability:
```suggestion
Another common task in remote systems is configuration management. For example, updating configuration files, pushing files to servers, pulling files generated by running services, and so on.
```

Then create our website _locally_:

```bash
echo "Hello, Pebble!" > /tmp/index.html
```

Create a directory in the remote system:

```bash
pebble mkdir -p /var/www/pebble
```

Push our local website file to the remote system:

```bash
pebble push /tmp/index.html /var/www/pebble/index.html
```

Create a virtual host configuration _locally_:

```
cat <<EOF > /tmp/pebble
server {
       listen 81;
       listen [::]:81;

       server_name example.ubuntu.com;

       root /var/www/pebble;
       index index.html;

       location / {
               try_files \$uri \$uri/ =404;
       }
}
EOF
```

Push the file to the remote system:

```bash
pebble push /tmp/pebble /etc/nginx/sites-enabled/pebble
```

Activate the newly added virtual host in the remote system by restarting Nginx:

```bash
pebble exec service nginx restart
```

Finally, we can test the final result:

```{terminal}
:input: curl localhost:81
Hello, Pebble!
```

For more information on files related commands in Pebble, see:

- {ref}`reference_pebble_ls_command`
- {ref}`reference_pebble_mkdir_command`
- {ref}`reference_pebble_rm_command`
- {ref}`reference_pebble_push_command`
- {ref}`reference_pebble_pull_command`

---

> ## Use Pebble APIs

Reviewer: Small suggestion for consistency with how we're referring to the API in other places:
```suggestion
## Use the API
```

> In previous sections, we used Pebble commands to execute commands and manage files. For automation purposes, we can also do the same with Pebble API.

Reviewer: Small suggestions for readability & consistency:
```suggestion
In the previous sections, we used Pebble commands to run commands and manage files. When automating these tasks, we can use the HTTP API that Pebble provides.
```

For example, to push a file to the remote system, we can POST to the `/v1/files` endpoint:

```bash
curl --unix-socket $PEBBLE/.pebble.socket -XPOST localhost:4000/v1/files -H "Content-Type: multipart/form-data" -F request='{"action": "write", "files": [{"path": "/var/www/pebble/index.html", "make-dirs": true, "permissions": "644"}]}' -F 'files=@/tmp/index.html;type=application/octet-stream;filename=/var/www/pebble/index.html'
```

---

## See more

> In this guide, we used Pebble commands and curl to access the APIs. We can also access Pebble API using the Go client and Python client. For more information, see [How to use the Pebble API to manage services](/how-to/use-the-pebble-api).

Reviewer: Small suggestion for readability & consistency:
```suggestion
In this guide, we used Pebble commands and curl to access the API. We can also access the API using Go and Python. For more information, see [How to use the Pebble API to manage services](/how-to/use-the-pebble-api).
```
